File: apps/api/cutpilot/schemas/creator.py
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import Field

from cutpilot.schemas.common import APIModel


class HighlightOut(APIModel):
    id: uuid.UUID
    asset_id: uuid.UUID | None
    start: float
    end: float
    title: str
    reason: str
    caption_suggestion: str
    score: float
    factors: dict[str, Any]
    category: str | None
    created_at: datetime


class HighlightRequest(APIModel):
    asset_id: uuid.UUID | None = None
    count: int = Field(default=5, ge=1, le=12)
    min_seconds: float = Field(default=20, ge=5, le=300)
    max_seconds: float = Field(default=60, ge=10, le=600)
    platform: str | None = None


class ShortsRequest(APIModel):
    count: int = Field(default=3, ge=1, le=10)
    duration: Literal[15, 30, 45, 60, 90] = 45
    platform: str | None = None
    asset_id: uuid.UUID | None = None
    highlight_ids: list[uuid.UUID] | None = None
    caption_preset: str = "bold"
    reframe: bool = True
    auto_render: bool = False
    render_preset: str | None = (
        None  # defaults to the platform preset (youtube_shorts | instagram_reel | tiktok)
    )


class ThumbnailCandidateOut(APIModel):
    asset_id: uuid.UUID
    time: float
    score: float
    factors: dict[str, Any]
    face: dict[str, Any] | None = None
    url: str


class ThumbnailsOut(APIModel):
    source_asset_id: uuid.UUID | None
    candidates: list[ThumbnailCandidateOut]


class BrollMatchOut(APIModel):
    asset_id: uuid.UUID
    filename: str
    score: float
    duration: float | None
    media_type: str
    reason: str


class BrollResolveResult(APIModel):
    placed: list[dict[str, Any]]
    unresolved: list[dict[str, Any]]
    state: dict[str, Any] | None = None


class BrollPlaceRequest(APIModel):
    marker_id: str
    asset_id: uuid.UUID
    duration: float | None = None
    timeline_id: uuid.UUID | None = None


class MulticamSyncRequest(APIModel):
    reference_asset_id: uuid.UUID
    asset_ids: list[uuid.UUID] = Field(min_length=1, max_length=8)
    place_on_timeline: bool = False
    timeline_id: uuid.UUID | None = None


class MulticamSyncOut(APIModel):
    reference_asset_id: uuid.UUID
    offsets: list[dict[str, Any]]
    state: dict[str, Any] | None = None


class SequenceSettingsUpdate(APIModel):
    caption_style: dict[str, Any] | None = None
    audio: dict[str, Any] | None = None
    aspect_ratio: str | None = None
    reframe_mode: Literal["center", "track", "off"] | None = None
    timeline_id: uuid.UUID | None = None

File: apps/web/lib/creator.ts
"use client";
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { apiGet, apiPatch, apiPost } from "@/lib/api";
import type { Highlight, Job } from "@/lib/types";
import { useJobsStore } from "@/stores/jobs";

export interface ThumbnailCandidate {
  asset_id: string;
  time: number;
  score: number;
  factors: Record<string, number>;
  face: { x: number; y: number; w: number; h: number } | null;
  url: string;
}

export function useHighlights(projectId: string | undefined) {
  return useQuery({ queryKey: ["highlights", projectId], queryFn: () => apiGet<Highlight[]>(`/projects/${projectId}/highlights`), enabled: !!projectId });
}

export function useThumbnails(projectId: string | undefined) {
  return useQuery({ queryKey: ["thumbnails", projectId], queryFn: () => apiGet<{ source_asset_id: string | null; candidates: ThumbnailCandidate[] }>(`/projects/${projectId}/thumbnails`), enabled: !!projectId });
}

export function useCreatorMutations(projectId: string) {
  const qc = useQueryClient();
  const upsertMany = useJobsStore((s) => s.upsertMany);
  const track = (res: { jobs: Job[] }) => upsertMany(res.jobs);
  const detectHighlights = useMutation({ mutationFn: (body: { count?: number; min_seconds?: number; max_seconds?: number; asset_id?: string | null }) => apiPost<{ jobs: Job[] }>(`/projects/${projectId}/highlights`, body), onSuccess: track });
  const generateShorts = useMutation({ mutationFn: (body: { count?: number; duration: 15 | 30 | 45 | 60 | 90; platform?: string; caption_preset?: string; reframe?: boolean; highlight_ids?: string[]; asset_id?: string | null; auto_render?: boolean; render_preset?: string | null }) => apiPost<{ jobs: Job[] }>(`/projects/${projectId}/shorts`, body), onSuccess: track });
  const generateThumbnails = useMutation({ mutationFn: () => apiPost<{ jobs: Job[] }>(`/projects/${projectId}/thumbnails`), onSuccess: track });
  const trackReframe = useMutation({ mutationFn: () => apiPost<{ jobs: Job[] }>(`/projects/${projectId}/reframe/track`), onSuccess: track });
  const resolveBroll = useMutation({
    mutationFn: (timelineId?: string | null) => apiPost<{ placed: unknown[]; unresolved: { marker_id: string; query: string }[] }>(`/projects/${projectId}/broll/resolve`, undefined, { timeline_id: timelineId ?? undefined }),
    onSuccess: () => qc.invalidateQueries({ queryKey: ["timeline", projectId] }),
  });
  const updateSequence = useMutation({
    mutationFn: (body: { caption_style?: Record<string, unknown>; audio?: Record<string, unknown>; aspect_ratio?: string; reframe_mode?: "center" | "track" | "off"; timeline_id?: string | null }) => apiPatch<{ version: number; document: unknown }>(`/projects/${projectId}/sequence`, body),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["timeline", projectId] });
      qc.invalidateQueries({ queryKey: ["timeline-versions", projectId] });
    },
  });
  const multicamSync = useMutation({
    mutationFn: (body: { reference_asset_id: string; asset_ids: string[]; place_on_timeline: boolean }) => apiPost<{ offsets: { asset_id: string; filename: string; offset: number; confidence: number }[] }>(`/projects/${projectId}/multicam/sync`, body),
    onSuccess: () => qc.invalidateQueries({ queryKey: ["timeline", projectId] }),
  });
  return { detectHighlights, generateShorts, generateThumbnails, trackReframe, resolveBroll, updateSequence, multicamSync };
}

export function useAIUsage(projectId: string | undefined) {
  return useQuery({
    queryKey: ["ai-usage", projectId],
    queryFn: () => apiGet<{ total_cost_usd: number; total_requests: number; items: { provider: string; model: string; operation: string; requests: number; estimated_cost_usd: number }[] }>(`/projects/${projectId}/ai-usage`),
    enabled: !!projectId,
    staleTime: 30_000,
  });
}
